presto-main-base/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ConvertHashJoinToSortMergeJoin.java

```diff
@@ -0,0 +1,99 @@
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.spi.plan.JoinNode;
+import com.facebook.presto.spi.plan.MergeJoinNode;
+import com.facebook.presto.spi.plan.Ordering;
+import com.facebook.presto.spi.plan.OrderingScheme;
+import com.facebook.presto.spi.plan.PlanNode;
+import com.facebook.presto.spi.plan.SortNode;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Optional;
+
+import static com.facebook.presto.SystemSessionProperties.preferSortMergeJoin;
+import static com.facebook.presto.common.block.SortOrder.ASC_NULLS_FIRST;
+import static com.facebook.presto.spi.plan.JoinDistributionType.PARTITIONED;
+import static com.facebook.presto.spi.plan.JoinDistributionType.REPLICATED;
+import static com.facebook.presto.sql.planner.plan.Patterns.join;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+public class ConvertHashJoinToSortMergeJoin
+        implements Rule<JoinNode>
+{
+    private static final Pattern<JoinNode> PATTERN = join().matching(
+            joinNode -> MergeJoinNode.isMergeJoinEligible(joinNode) &&
+                    // it's not worth doing a sort merge join for a broadcast join which will have a small build side
+                    !joinNode.getDistributionType().orElse(PARTITIONED).equals(REPLICATED));
+
+    @Override
+    public Pattern<JoinNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return preferSortMergeJoin(session);
+    }
+
+    @Override
+    public Result apply(JoinNode node, Captures captures, Context context)
+    {
+        PlanNode left = node.getLeft();
+        PlanNode right = node.getRight();
+
+        List<Ordering> leftOrdering = node.getCriteria().stream()
+                .map(criterion -> new Ordering(criterion.getLeft(), ASC_NULLS_FIRST))
+                .collect(toImmutableList());
+        PlanNode sortedLeft = new SortNode(
+                Optional.empty(),
+                context.getIdAllocator().getNextId(),
+                left,
+                new OrderingScheme(leftOrdering),
+                true,
+                ImmutableList.of());
+
+        List<Ordering> rightOrdering = node.getCriteria().stream()
+                .map(criterion -> new Ordering(criterion.getRight(), ASC_NULLS_FIRST))
+                .collect(toImmutableList());
+        PlanNode sortedRight = new SortNode(
+                Optional.empty(),
+                context.getIdAllocator().getNextId(),
+                right,
+                new OrderingScheme(rightOrdering),
+                true,
+                ImmutableList.of());
+
+        return Result.ofPlanNode(
+                new MergeJoinNode(
+                        Optional.empty(),
+                        node.getId(),
+                        node.getType(),
+                        sortedLeft,
```

Comment:
This seems to be always adding the `SortNode`. We should add SortNode conditionally. 

You can use `PropertyDerivations.derivePropertiesRecursively()` to obtain the properties and check for `SortingProperty`s to match the JoinKey. If its present you can skip creating this sort node.

Reply:
Thanks for pointing this out. Maybe we can keep this version minimal and iterate on followup PRs.